File: app/views/login_view.py
from flet import *
from app.controllers.user_controller import UserController
from navigation_view import navigate_to, create_navigation_drawer


class UserWidget(UserControl):
    def __init__(self, title: str, sub_title: str, btn_name: str, link: str, forgot_password: str,
                 navigate_to_registration):
        super().__init__()
        self.title = title
        self.sub_title = sub_title
        self.btn_name = btn_name
        self.link = link
        self.forgot_password = forgot_password
        self.navigate_to_registration = navigate_to_registration
        self.user_controller = UserController()
        super().__init__()

    def InputTextField(self, text: str, hide: bool, ref):
        return Container(
            alignment=alignment.center,
            content=TextField(
                height=58,
                width=300,
                bgcolor="#f0f3f6",
                text_size=14,
                color="black",
                border_color="transparent",
                hint_text=text,
                filled=True,
                cursor_color="black",
                hint_style=TextStyle(
                    size=13,
                    color="black",
                ),
                password=hide,
                ref=ref,
            ),
        )

    def SignInOption(self, path: str, name: str):
        return Container(
            alignment=alignment.center,
            content=ElevatedButton(
                content=Row(
                    alignment="center",
                    spacing=4,
                    controls=[
                        Image(
                            src=path,
                            width=35,
                            height=35,
                        ),
                        Text(
                            name,
                            color="black",
                            size=12,
                            weight="bold",
                        )
                    ],
                ),
                style=ButtonStyle(
                    shape={
                        "": RoundedRectangleBorder(radius=8),
                    },
                    bgcolor={
                        "": "#f0f3f6",
                    }
                ),
            ),
        )

    def show_dialog(self, page, title, content, color, on_close):
        dlg_modal = AlertDialog(
            modal=True,
            title=Text(title, color=color),
            content=Text(content),
            actions=[
                TextButton("OK", on_click=lambda e: [page.close(dlg_modal), on_close(e)])
            ],
            actions_alignment=MainAxisAlignment.END
        )
        page.open(dlg_modal)

    def login(self, e):
        email = self.email_input.current.value
        password = self.password_input.current.value

        result = self.user_controller.login_user(email, password)

        if result["status"] == "error":
            self.show_dialog(e.page, "Error", result["message"], colors.RED, lambda _: None)
        else:
            self.email_input.current.value = ""
            self.password_input.current.value = ""
            self.show_dialog(e.page, "Success", result["message"], colors.GREEN, lambda e: navigate_to(e.page, "Register"))

    def build(self):
        self.email_input = Ref[TextField]()
        self.password_input = Ref[TextField]()

        self._title = Container(
            alignment=alignment.center,
            content=Text(
                self.title,
                size=24,
                text_align="center",
                weight="bold",
                color="black",
            )
        )

        self._sub_title = Container(
            alignment=alignment.center,
            content=Text(
                self.sub_title,
                size=18,
                text_align="center",
                color="black",
            )
        )

        self._link = Container(
            alignment=alignment.center,
            content=Row(
                alignment="center",
                controls=[
                    Text("Don't have an account?", size=12, color="black"),
                    Container(
                        content=Text(
                            self.link,
                            size=12,
                            color="green",
                            weight="bold",
                        ),
                        on_click=self.navigate_to_registration,
                    )
                ]
            )
        )

        self._forgot_password = Container(
            alignment=alignment.center_right,
            content=Text(
                self.forgot_password,
                size=12,
                color="green",
                weight="bold",
            )
        )

        self._sign_in = Container(
            alignment=alignment.center,
            content=ElevatedButton(
                on_click=self.login,
                content=Text(
                    self.btn_name,
                    size=14,
                    weight="bold",
                ),
                bgcolor="black",
                color="white",
                style=ButtonStyle(
                    shape={
                        "": RoundedRectangleBorder(radius=8)
                    },
                ),
                height=58,
                width=300,
            )
        )

        class DividerWithText(Container):
            def __init__(self, text):
                super().__init__(
                    alignment=alignment.center,
                    content=Row(
                        alignment='center',
                        controls=[
                            Container(width=120, height=1, bgcolor="black"),
                            Text(text, color="black", weight='w400'),
                            Container(width=120, height=1, bgcolor="black"),
                        ],
                        spacing=10,
                    )
                )

        return Column(
            horizontal_alignment="center",
            alignment="center",
            controls=[
                self._title,
                self._sub_title,
                Container(padding=5),
                Column(
                    spacing=15,
                    horizontal_alignment="center",
                    controls=[
                        self.InputTextField("Email", False, self.email_input),
                        self.InputTextField("Password", True, self.password_input),
                        self._forgot_password,
                    ],
                ),
                Container(padding=3),
                self._sign_in,
                Container(padding=3),
                Column(
                    horizontal_alignment="center",
                    controls=[
                        DividerWithText("or"),
                        self.SignInOption("../assets/icon.png", "Facebook"),
                        self.SignInOption("../assets/icon.png", "Google"),
                    ],
                ),
                Container(padding=5),
                self._link,
            ],
        )


def login_page(page: Page):
    def _main_column():
        return Container(
            alignment=alignment.center,
            width=320,
            height=650,
            bgcolor="#ffffff",
            padding=16,
            border_radius=35,
            content=Column(
                spacing=25,
                horizontal_alignment="center",
                alignment="center",
            )
        )

    _sign_in_ = UserWidget(
        "Welcome Back!",
        "Enter your account details below",
        "Sign In",
        "Sign Up",
        "Forgot Password?",
        lambda e: navigate_to(page, "Register")
    )

    _sign_in_main = _main_column()
    _sign_in_main.content.controls.append(Container(padding=15))
    _sign_in_main.content.controls.append(_sign_in_)

    drawer = create_navigation_drawer(page)
    page.add(
        AppBar(
            Row(
                controls=[
                    IconButton(
                        icon=icons.MENU_ROUNDED,
                        icon_size=25,
                        icon_color="white",
                        on_click=lambda e: page.open(drawer),
                    ),
                ],
            ),
            title=Text('Login', color="white"),
            bgcolor="black",
        ),
        Container(
            alignment=alignment.center,
            expand=True,
            content=Column(
                alignment=alignment.center,
                horizontal_alignment="center",
                controls=[
                    _sign_in_main,
                ]
            )
        )
    )
    page.update()

File: frontend/lib/view/login_view.dart
import 'package:flutter/material.dart';
import 'package:font_awesome_flutter/font_awesome_flutter.dart';
import 'package:frontend/view/dashboard_view.dart';
import 'package:frontend/view/registration_view.dart';

class UserWidget extends StatefulWidget {
  final String title;
  final String subTitle;
  final String btnName;
  final String link;
  final String forgotPassword;
  final VoidCallback navigateToRegistration;

  const UserWidget({
    super.key,
    required this.title,
    required this.subTitle,
    required this.btnName,
    required this.link,
    required this.forgotPassword,
    required this.navigateToRegistration,
  });

  @override
  _UserWidgetState createState() => _UserWidgetState();
}

class _UserWidgetState extends State<UserWidget> {
  final TextEditingController identifierController = TextEditingController();
  final TextEditingController passwordController = TextEditingController();
  bool isRemembered = false;

  void login() {
    final identifier = identifierController.text;
    final password = passwordController.text;

    if (identifier.isEmpty || password.isEmpty) {
      showDialog(
        context: context,
        builder: (context) => AlertDialog(
          title: const Text('Error'),
          content: const Text('Please fill in all fields'),
          actions: [
            TextButton(
              onPressed: () => Navigator.pop(context),
              child: const Text('OK'),
            ),
          ],
        ),
      );
    } else {
      Navigator.push(
        context,
        MaterialPageRoute(
          builder: (context) => const DashboardView(),
        ),
      );
    }
  }

  Widget inputTextField(
      String text, bool hide, TextEditingController controller) {
    return TextField(
      controller: controller,
      obscureText: hide,
      decoration: InputDecoration(
        hintText: text,
        filled: true,
        fillColor: Colors.grey.shade200,
        border: OutlineInputBorder(
          borderRadius: BorderRadius.circular(8),
          borderSide: BorderSide.none,
        ),
      ),
    );
  }

  Widget socialButton(IconData icon, Color color) {
    return ElevatedButton(
      onPressed: () {},
      style: ElevatedButton.styleFrom(
        padding: const EdgeInsets.all(16),
        backgroundColor: Colors.grey.shade200,
        shape: const CircleBorder(),
      ),
      child: FaIcon(icon, size: 30, color: color),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SingleChildScrollView(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            const SizedBox(height: 20),
            Center(
              child: Image.asset(
                'assets/img/test.png',
                height: 150,
                fit: BoxFit.contain,
              ),
            ),
            const SizedBox(height: 20),
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: 20),
              child: Column(
                children: [
                  Text(
                    widget.title,
                    style: const TextStyle(
                        fontSize: 24, fontWeight: FontWeight.bold),
                    textAlign: TextAlign.center,
                  ),
                  const SizedBox(height: 10),
                  Text(
                    widget.subTitle,
                    style: const TextStyle(fontSize: 18),
                    textAlign: TextAlign.center,
                  ),
                  const SizedBox(height: 20),
                  inputTextField(
                      'Email or Username', false, identifierController),
                  const SizedBox(height: 10),
                  inputTextField('Password', true, passwordController),
                  const SizedBox(height: 10),
                  Row(
                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                    children: [
                      Row(
                        children: [
                          Checkbox(
                            value: isRemembered,
                            onChanged: (bool? newValue) {
                              setState(() {
                                isRemembered = newValue ?? false;
                              });
                            },
                          ),
                          const Text("Remember me"),
                        ],
                      ),
                      TextButton(
                        onPressed: () {},
                        child: Text(
                          widget.forgotPassword,
                          style: const TextStyle(
                              color: Colors.green, fontWeight: FontWeight.bold),
                        ),
                      ),
                    ],
                  ),
                  const SizedBox(height: 10),
                  SizedBox(
                    width: double.infinity,
                    child: ElevatedButton(
                      onPressed: login,
                      style: ElevatedButton.styleFrom(
                        fixedSize: const Size.fromHeight(58),
                        backgroundColor: Colors.black,
                        shape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(8),
                        ),
                      ),
                      child: Text(
                        widget.btnName,
                        style: const TextStyle(
                            fontSize: 16,
                            fontWeight: FontWeight.bold,
                            color: Colors.white),
                      ),
                    ),
                  ),
                  const SizedBox(height: 20),
                  Row(
                    children: [
                      Expanded(child: Divider(color: Colors.black)),
                      const SizedBox(width: 10),
                      const Text('or',
                          style: TextStyle(fontWeight: FontWeight.w400)),
                      const SizedBox(width: 10),
                      Expanded(child: Divider(color: Colors.black)),
                    ],
                  ),
                  const SizedBox(height: 20),
                  Row(
                    mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                    children: [
                      socialButton(FontAwesomeIcons.facebookF, Color(0xFF3b5998)),
                      socialButton(FontAwesomeIcons.google, Colors.red),
                      socialButton(FontAwesomeIcons.apple, Colors.black),
                    ],
                  ),
                  const SizedBox(height: 20),
                  Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      const Text("Don't have an account?",
                          style: TextStyle(fontSize: 12)),
                      TextButton(
                        onPressed: () {
                          Navigator.push(
                            context,
                            MaterialPageRoute(
                              builder: (context) => const RegistrationView(), 
                            ),
                          );
                        },
                        child: Text(
                          widget.link,
                          style: const TextStyle(
                              color: Colors.green, fontWeight: FontWeight.bold),
                        ),
                      ),
                    ],
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class LoginView extends StatelessWidget {
  const LoginView({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: UserWidget(
        title: 'Welcome!',
        subTitle: 'Enter your account details below',
        btnName: 'Sign In',
        link: 'Sign Up',
        forgotPassword: 'Forgot Password?',
        navigateToRegistration: () {
          Navigator.push(
            context,
            MaterialPageRoute(
              builder: (context) => const RegistrationView(),
            ),
          );
        },
      ),
    );
  }
}
